tests/integration/cmd_namespace_test.go

```diff
@@ -2,87 +2,91 @@ package integration
+				By("using an existent name", func() {
+					helper.Cmd("odo", "create", commandName, commonVar.Project).ShouldFail()
+				})
+				By("using an invalid name", func() {
+					helper.Cmd("odo", "create", commandName, "12345").ShouldFail()
+					Expect(commonVar.CliRunner.GetActiveNamespace()).To(Equal(commonVar.Project))
+				})
+			})
 		})
-	})
 
-	for _, commandName := range []string{"namespace", "project"} {
-		When(fmt.Sprintf("using the alias %[1]s to delete a %[1]s", commandName), func() {
-			var namespace string
+		Describe("delete "+commandName, func() {
 
-			BeforeEach(func() {
-				namespace = helper.CreateRandProject()
-				Expect(commonVar.CliRunner.CheckNamespaceProjectExists(namespace)).To(BeTrue())
-			})
+			When("force-deleting a valid "+commandName, func() {
+				var namespace string
 
-			checkNsDeletionFunc := func(wait bool) {
-				args := []string{"delete", commandName, namespace, "--force"}
-				if wait {
-					args = append(args, "--wait")
-				}
-				out := helper.Cmd("odo", args...).ShouldPass().Out()
-				if wait {
-					Expect(commonVar.CliRunner.GetAllNamespaceProjects()).ShouldNot(ContainElement(namespace))
-				} else {
-					Eventually(func() []string {
-						return commonVar.CliRunner.GetAllNamespaceProjects()
-					}, 60*time.Second).ShouldNot(ContainElement(namespace))
+				BeforeEach(func() {
+					namespace = helper.CreateRandProject()
+					Expect(commonVar.CliRunner.CheckNamespaceProjectExists(namespace)).To(BeTrue())
+				})
+
+				checkNsDeletionFunc := func(additionalArgs []string, nsCheckerFunc func()) {
+					args := []string{"delete", commandName, namespace, "--force"}
+					if additionalArgs != nil {
+						args = append(args, additionalArgs...)
+					}
+					out := helper.Cmd("odo", args...).ShouldPass().Out()
+					if nsCheckerFunc != nil {
+						nsCheckerFunc()
+					}
+					Expect(out).To(
+						ContainSubstring(fmt.Sprintf("%s %q deleted", strings.Title(commandName), namespace)))
 				}
-				Expect(out).To(
-					ContainSubstring(fmt.Sprintf("%s %q deleted", strings.Title(commandName), namespace)))
-			}
 
-			It(fmt.Sprintf("should successfully delete the %s using the force flag and asynchronously", commandName), func() {
-				checkNsDeletionFunc(false)
-			})
+				It(fmt.Sprintf("should successfully delete the %s asynchronously", commandName), func() {
+					checkNsDeletionFunc(nil, func() {
+						Eventually(func() []string {
+							return commonVar.CliRunner.GetAllNamespaceProjects()
+						}, 60*time.Second).ShouldNot(ContainElement(namespace))
+					})
+				})
 
-			It(fmt.Sprintf("should successfully delete the %s using the force flag and waiting", commandName), func() {
-				checkNsDeletionFunc(true)
+				It(fmt.Sprintf("should successfully delete the %s synchronously with --wait", commandName), func() {
+					checkNsDeletionFunc([]string{"--wait"}, func() {
+						Expect(commonVar.CliRunner.GetAllNamespaceProjects()).ShouldNot(ContainElement(namespace))
```

Comment:
Could you try to use `kubectl get namespace <name>` instead of the list?

We are experiencing flaky errors where the namespace is in the list. Is it due to some cache or some delay in the list update?

Reply:
Ok. I remember I tried to use one of the existing methods (one of `helper.GetNamespaceProject` or `CheckNamespaceProjectExists`) but couldn't because they expect that the "kubectl get namespace &lt;name&gt;" command passes. So I couldn't use it to check that a namespace no longer exists. That's why I tried with the list.
But let me fix that.

Reply:
> We are experiencing flaky errors where the namespace is in the list. Is it due to some cache or some delay in the list update?

I quickly tested on a `cluster-bot` cluster, and from what I understand from the API Server response headers, `kubectl` could indeed cache responses, but the response must be validated with the origin server before each reuse. So to me, on the `kubectl` client side, responses are potentially cached but always checked for content updates.

```
❯ kubectl get ns --v=8
Alias tip: k get ns --v=8
I0516 13:41:26.906356  420945 loader.go:372] Config loaded from file:  /home/asoro/work/tmp/cluster-bot/config
I0516 13:41:26.908921  420945 round_trippers.go:463] GET https://127.0.0.1:43051/api/v1/namespaces?limit=500
I0516 13:41:26.908931  420945 round_trippers.go:469] Request Headers:
I0516 13:41:26.908943  420945 round_trippers.go:473]     Accept: application/json;as=Table;v=v1;g=meta.k8s.io,application/json;as=Table;v=v1beta1;g=meta.k8s.io,application/json
I0516 13:41:26.908951  420945 round_trippers.go:473]     User-Agent: kubectl/v1.24.0 (linux/amd64) kubernetes/4ce5a89
I0516 13:41:26.914511  420945 round_trippers.go:574] Response Status: 200 OK in 5 milliseconds
I0516 13:41:26.914524  420945 round_trippers.go:577] Response Headers:
I0516 13:41:26.914534  420945 round_trippers.go:580]     Audit-Id: 8dcfdfbf-47b7-40aa-a89e-e4547f653c8d
I0516 13:41:26.914544  420945 round_trippers.go:580]     Cache-Control: no-cache, private
I0516 13:41:26.914554  420945 round_trippers.go:580]     Content-Type: application/json
I0516 13:41:26.914561  420945 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: 915fbd93-565d-4fc2-8975-4dae2b4dcbea
I0516 13:41:26.914568  420945 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 98f3f965-3d3d-44fc-99d8-d6428ea1ffb3
I0516 13:41:26.914576  420945 round_trippers.go:580]     Date: Mon, 16 May 2022 11:41:26 GMT
```

I don't know how the `--wait` flag is used by `odo` under the hood, but maybe we are not guaranteed that the namespace is really deleted when the `project.Delete` method returns?

Anyways, I am trying to rely on the `kubectl get namespace <name>` command instead.

Reply:
Another cause would be that, in the sources (kclient/project.go), we are waiting for the deletion of the project, not the namespace